### ISSUE TITLE
Different initial loss between streaming and non-streaming pretraining

### ISSUE THREAD

**Issue by LucasCoutoLima** [axolotl-ai-cloud/axolotl#3441]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

When using the same dataset, seed, and deterministic settings, the initial training loss during pretraining should be identical (or extremely close) between:

pretraining_dataset: type: pretrain (streaming-style pipeline)

datasets: type: completion (non-streaming pipeline)

Specifically, the first training step should receive exactly the same tokenized data in both cases when no samples are dropped and shuffling is disabled.

### Current behaviour

The initial loss differs significantly between the two configurations, even though deterministic settings are used.

Observed:

With pretraining_dataset: type: pretrain → initial loss ≈ 2.859

With datasets: type: completion → initial loss ≈ 2.125
Configuration used:

task_type: pretraining
pretraining_dataset:
  path: dataset_path
  ds_type: parquet
  text_column: text
  type: pretrain

seed: 42
pretrain_multipack_buffer_size: 100000
group_by_length: false
shuffle_merged_datasets: false
sample_packing_sequentially: true
curriculum_sampling: true

Notes:
No samples appear to be dropped.
Shuffling is disabled.
A large multipack buffer is used to try to ensure identical first-step data.

### Steps to reproduce

1. Run pretraining with streaming-style config:
task_type: pretraining
pretraining_dataset:
  path: dataset_path
  ds_type: parquet
  text_column: text
  type: pretrain

seed: 42
pretrain_multipack_buffer_size: 100000
group_by_length: false
shuffle_merged_datasets: false
sample_packing_sequentially: true
curriculum_sampling: true

2. Run pretraining again using the non-streaming dataset config:
task_type: pretraining
datasets:
  - path: dataset_path
    ds_type: parquet
    text_column: text
    type: completion

seed: 42
group_by_length: false
shuffle_merged_datasets: false
sample_packing_sequentially: true
curriculum_sampling: true

3. Compare the first-step losses

### Config yaml

```yaml
base_model: Qwen/Qwen2.5-0.5B-Instruct
model_type: AutoModelForCausalLM
tokenizer_type: AutoTokenizer
bf16: true
# Dataset
task_type: pretraining
datasets: #pretraining_dataset:
  - path: dataset_path
    ds_type: parquet
    text_column: text
    type: completion #pretrain
seed: 42
sample_packing_bin_size: 100000
# streaming: true
# streaming_multipack_buffer_size: 10000
sequence_len: 32768
micro_batch_size: 24
gradient_accumulation_steps: 1
max_steps: 1
sample_packing: true
flash_attention: true
pretrain_multipack_attn: true
group_by_length: false
shuffle_merged_datasets: false
sample_packing_sequentially: true
curriculum_sampling: true
logging_steps: 1
optimizer: adamw_bnb_8bit
learning_rate: 1e-5
lr_scheduler: cosine
cosine_min_lr_ratio: 0.1
weight_decay: 0.0
warmup_steps: 0   
output_dir: test_loss/
plugins:
  - axolotl.integrations.liger.LigerPlugin
liger_rope: true
liger_rms_norm: false
liger_glu_activation: true
liger_layer_norm: true
liger_fused_linear_cross_entropy: true
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
```

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12.11

### axolotl branch-commit

0.15.0.dev0

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by sumukhacharya03** [comment]
"Hey, I'd like to work on this. Can I be assigned?"
